README
= MerbRecaptcha

http://github.com/felix/merb_recaptcha/tree/master

== DESCRIPTION:

This is a Merb 'take' on the original code by McClain Looney http://www.loonsoft.com/recaptcha/.

== FEATURES/PROBLEMS:

Still need to test this!

== SYNOPSIS:

Enables easy integration of the ReCaptcha technique into your Merb application. 
For more information on ReCaptcha see http://recaptcha.net/

== REQUIREMENTS:

Not sure yet.

== INSTALL:

First, create an account at ReCaptcha.net at http://www.recaptcha.net/

Get your keys, and make them available as constants in your application. You can do this however you want, but RCC_PUB, RCC_PRIV (for regular reCaptcha) and MH_PUB MH_PRIV (for MailHide) must be set to their respective values (the keys you receive from reCaptcha).

The common methods of doing this are to set these variables in your environment/production.rb file, or via an environment variable.

The ReCaptchaClient initializer can also take an options hash containing keys thusly:

  RecaptchaClient.new(:rcc_pub=>'some key', :rcc_priv=>'some other key')

After your keys are configured, require and include the gem into your init.rb:

dependency "merb_recaptcha"

then include the helper into your controller:

class User < Application
  include MerbRecaptcha::AppHelper

This will mix in validate_recap to your application controller. Then, in your concrete controller, chain validate_recap into your validation:

def create
  @user = User.new(params[:user])
    if validate_recap(params, @user.errors) && @user.save
      ...do stuff...

Require and include the view helper in the appropriate helper:

module Merb
  module UserHelper
    include MerbRecaptcha::ViewHelper
  end
end


This will mix in get_captcha().

Now you can just call

get_captcha()

in your view to insert the requisite widget from ReCaptcha.

To customize theme and tabindex of the widget, you can include an options hash:

get_captcha(:options => {:theme => 'white', :tabindex => 10})

See the reCAPTCHA API Documentation under “Look and Feel Customization” for more information.
Proxy support
If your rails application requires the use of a proxy, set proxy_host into your environment:

  ENV['proxy_host']='foo.example.com:8080'

Mail Hide

When you mix in ViewHelper as above, you also get

 mail_hide(address, contents)

, which you can call in your view thusly:

  ...
  <%= mail_hide(user.email) %>

Contents defaults to the first few characters of the email address.


== LICENSE:

(The MIT License)

Copyright (c) 2008 Felix McCoey

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


HISTORY
== 0.0.1 2008-06-09

* 1 mahor enhancement
  * Initial release


test/test_merb_recaptcha.rb
require File.dirname(__FILE__) + '/test_helper.rb'
require 'rubygems'
gem 'mocha'
require 'mocha'
gem 'rails'

module MerbRecaptcha
  module ViewHelper
    def self.define_public_key
      class_eval "RCC_PUB = 'foo'"
    end
    def self.define_private_key
      class_eval "RCC_PRIV = 'bar'"
    end
    def self.undefine_public_key
      remove_const :RCC_PUB
    end
    def self.undefine_private_key
      remove_const :RCC_PRIV
    end
  end
  module AppHelper
    def self.define_public_key
      class_eval "RCC_PUB = 'foo'"
    end
    def self.define_private_key
      class_eval "RCC_PRIV = 'bar'"
    end
    def self.undefine_public_key
      remove_const :RCC_PUB
    end
    def self.undefine_private_key
      remove_const :RCC_PRIV
    end
    public :validate_recap
  end
end

class TestMerbRecaptcha < Test::Unit::TestCase
  class ViewFixture
    PRIVKEY='6LdnAQAAAAAAAPYLVPwVvR7Cy9YLhRQcM3NWsK_C'
    PUBKEY='6LdnAQAAAAAAAKEk59yjuP3csGEeDmdj__7cyMtY'
    include MerbRecaptcha::ViewHelper
    # views in Rails "inherit" the controller's session...
    def session
      @session ||= {}
    end
  end
  class ControllerFixture
    include MerbRecaptcha::AppHelper
    # all Rails controllers have a session...
    def session
      @session ||= {}
    end
    include Mocha::Standalone
    def request
      request = mock()
      request.stubs(:remote_ip).returns('0.0.0.0') # this will skip the actual captcha validation...
      request
    end
  end

  def setup
    @vf = ViewFixture.new
    @cf = ControllerFixture.new
  end

  def test_encrypt
    mhc = MerbRecaptcha::MHClient.new('01S1TOX9aibKxfC9oJlp8IeA==', 'deadbeefdeadbeefdeadbeefdeadbeef')
    z =mhc.encrypt('x@example.com')
    assert_equal 'wBG7nOgntKqWeDpF9ucVNQ==', z
    z =mhc.encrypt('johndoe@example.com')
    assert_equal 'whWIqk0r4urZ-3S7y7uSceC9_ECd3hpAGy71E2o0HpI=', z
  end
  def test_encrypt_long
    mhc = MerbRecaptcha::MHClient.new('01S1TOX9aibKxfC9oJlp8IeA==', 'deadbeefdeadbeefdeadbeefdeadbeef')
    z =mhc.encrypt('averylongemailaddressofmorethan32cdharactersx@example.com')
    assert_equal "q-0LLVT2bIxWbFpfLfpNhJAGadkfWXVk4hAxSlVaLrdnXrsB1NKNubavS5N-7PBued3K531vifN6NB3iz3W7qQ==",z
    z =mhc.encrypt('johndoe@example.com')
    assert_equal 'whWIqk0r4urZ-3S7y7uSceC9_ECd3hpAGy71E2o0HpI=', z
  end

  def test_nil_challenge
    client = new_client
    estub = stub_everything('errors')
    client.validate('abc', nil, 'foo', estub)

  end

  def test_constructor
    client = new_client('abc', 'def', true)
    expected= <<-EOF
    <script type=\"text/javascript\" src=\"https://api-secure.recaptcha.net/challenge?k=abc\"> </script>\n      <noscript>\n      <iframe src=\"https://api-secure.recaptcha.net/noscript?k=abc\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    assert_equal expected.strip, client.get_challenge.strip
    client = new_client
    expected= <<-EOF
    <script type=\"text/javascript\" src=\"http://api.recaptcha.net/challenge?k=abc\"> </script>\n      <noscript>\n      <iframe src=\"http://api.recaptcha.net/noscript?k=abc\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    assert_equal expected.strip, client.get_challenge.strip
    client = new_client
    expected= <<-EOF
    <script type=\"text/javascript\" src=\"http://api.recaptcha.net/challenge?k=abc\"> </script>\n      <noscript>\n      <iframe src=\"http://api.recaptcha.net/noscript?k=abc\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    assert_equal expected.strip, client.get_challenge.strip
  end
  
  def test_constructor_with_recaptcha_options
    # "Look and Feel Customization" per http://recaptcha.net/apidocs/captcha/
    client = new_client
    expected= <<-EOF
    <script type=\"text/javascript\">\nvar RecaptchaOptions = { theme : \"white\", tabindex : 10};\n</script>\n      <script type=\"text/javascript\" src=\"http://api.recaptcha.net/challenge?k=abc&error=somerror\"> </script>\n      <noscript>\n      <iframe src=\"http://api.recaptcha.net/noscript?k=abc&error=somerror\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    assert_equal expected.strip, client.get_challenge('somerror', :options => {:theme => 'white', :tabindex => 10}).strip
  end

  def test_validate_fails
    badwords_resp="false\r\n360 incorrect-captcha-sol"
    err_stub=mock()
    err_stub.expects(:add_to_base).with("Captcha failed.")
    stub_proxy=mock('proxy')
    stub_http = mock('http mock')
    stub_proxy.expects(:start).with('api-verify.recaptcha.net').returns(stub_http)
    stub_http.expects(:post).with('/verify', 'privatekey=def&remoteip=localhost&challenge=abc&response=def', {'Content-Type' => 'application/x-www-form-urlencoded'}).returns(['foo', badwords_resp])
    Net::HTTP.expects(:Proxy).returns(stub_proxy)
    client = new_client
    assert !client.validate('localhost', 'abc', 'def', err_stub)
  end
  def test_validate_good
    goodwords_resp="true\r\nsuccess"
    err_stub=mock()
    stub_proxy=mock('proxy')
    stub_http = mock('http mock')
    stub_proxy.expects(:start).with('api-verify.recaptcha.net').returns(stub_http)
    stub_http.expects(:post).with('/verify', 'privatekey=def&remoteip=localhost&challenge=abc&response=def', {'Content-Type' => 'application/x-www-form-urlencoded'}).returns(['foo', goodwords_resp])
    Net::HTTP.expects(:Proxy).with(nil, nil).returns(stub_proxy)
    client = new_client
    assert client.validate('localhost', 'abc', 'def', err_stub)
  end
  def test_validate_good_proxy
    ENV['proxy_host']='fubar:8080'
    goodwords_resp="true\r\nsuccess"
    err_stub=mock()
    stub_proxy=mock('proxy')
    stub_http = mock('http mock')
    stub_proxy.expects(:start).with('api-verify.recaptcha.net').returns(stub_http)
    stub_http.expects(:post).with('/verify', 'privatekey=def&remoteip=localhost&challenge=abc&response=def', {'Content-Type' => 'application/x-www-form-urlencoded'}).returns(['foo', goodwords_resp])
    Net::HTTP.expects(:Proxy).with('fubar', '8080').returns(stub_proxy)
    client = new_client
    assert client.validate('localhost', 'abc', 'def', err_stub)
    ENV['proxy_host']='fubar'
    err_stub=mock()
    stub_proxy=mock('proxy')
    stub_http = mock('http mock')
    stub_proxy.expects(:start).with('api-verify.recaptcha.net').returns(stub_http)
    stub_http.expects(:post).with('/verify', 'privatekey=def&remoteip=localhost&challenge=abc&response=def', {'Content-Type' => 'application/x-www-form-urlencoded'}).returns(['foo', goodwords_resp])
    Net::HTTP.expects(:Proxy).with('fubar', nil).returns(stub_proxy)
    client = new_client
    assert client.validate('localhost', 'abc', 'def', err_stub)
  end
  
  #
  # unit tests for the get_captcha() ViewHelper method
  #
  
  def test_get_captcha_fails_without_key_constants
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PRIV)
    assert_raise NameError do
      @vf.get_captcha
    end
  end
  def test_get_captcha_fails_without_public_key_constant
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PRIV)
    MerbRecaptcha::ViewHelper.define_private_key
    assert MerbRecaptcha::ViewHelper.const_defined?(:RCC_PRIV)
    assert_raise NameError do
      @vf.get_captcha
    end
    MerbRecaptcha::ViewHelper.undefine_private_key
  end
  def test_get_captcha_fails_without_private_key_constant
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PRIV)
    MerbRecaptcha::ViewHelper.define_public_key
    assert MerbRecaptcha::ViewHelper.const_defined?(:RCC_PUB)
    assert_raise NameError do
      @vf.get_captcha
    end
    MerbRecaptcha::ViewHelper.undefine_public_key
  end
  def test_get_captcha_succeeds_with_key_constants
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PRIV)
    MerbRecaptcha::ViewHelper.define_public_key
    MerbRecaptcha::ViewHelper.define_private_key
    assert MerbRecaptcha::ViewHelper.const_defined?(:RCC_PUB)
    assert MerbRecaptcha::ViewHelper.const_defined?(:RCC_PRIV)
    assert_nothing_raised do
      @vf.get_captcha
    end
    MerbRecaptcha::ViewHelper.undefine_public_key
    MerbRecaptcha::ViewHelper.undefine_private_key
  end
  def test_get_captcha_succeeds_without_key_constants_but_with_options
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::ViewHelper.const_defined?(:RCC_PRIV)
    assert_nothing_raised do
      @vf.get_captcha(:rcc_pub => 'foo', :rcc_priv => 'bar')
    end
  end
  def test_get_captcha_is_correct_with_constants_and_with_options
    expected= <<-EOF
    <script type=\"text/javascript\" src=\"http://api.recaptcha.net/challenge?k=%s\"> </script>\n      <noscript>\n      <iframe src=\"http://api.recaptcha.net/noscript?k=%s\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    # first, with constants
    MerbRecaptcha::ViewHelper.define_public_key  # 'foo'
    MerbRecaptcha::ViewHelper.define_private_key # 'bar'
    actual = @vf.get_captcha
    assert_equal(((expected % ['foo', 'foo']).strip), actual.strip)
    MerbRecaptcha::ViewHelper.undefine_public_key
    MerbRecaptcha::ViewHelper.undefine_private_key
    # next, with options
    actual = @vf.get_captcha(:rcc_pub => 'foobar', :rcc_priv => 'blegga')
    assert_equal(((expected % ['foobar', 'foobar']).strip), actual.strip)
  end
  
  #
  # unit tests for the validate_recap() AppHelper method
  #
  
  def test_validate_recap_fails_without_key_constants
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PRIV)
    assert_raise NameError do
      @cf.validate_recap({}, {})
    end
  end
  def test_validate_recap_fails_without_public_key_constant
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PRIV)
    MerbRecaptcha::AppHelper.define_private_key
    assert MerbRecaptcha::AppHelper.const_defined?(:RCC_PRIV)
    assert_raise NameError do
      @cf.validate_recap({}, {})
    end
    MerbRecaptcha::AppHelper.undefine_private_key
  end
  def test_validate_recap_fails_without_private_key_constant
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PRIV)
    MerbRecaptcha::AppHelper.define_public_key
    assert MerbRecaptcha::AppHelper.const_defined?(:RCC_PUB)
    assert_raise NameError do
      @cf.validate_recap({}, {})
    end
    MerbRecaptcha::AppHelper.undefine_public_key
  end
  def test_validate_recap_succeeds_with_key_constants
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PRIV)
    MerbRecaptcha::AppHelper.define_public_key
    MerbRecaptcha::AppHelper.define_private_key
    assert MerbRecaptcha::AppHelper.const_defined?(:RCC_PUB)
    assert MerbRecaptcha::AppHelper.const_defined?(:RCC_PRIV)
    assert_nothing_raised do
      @cf.validate_recap({}, {})
    end
    MerbRecaptcha::AppHelper.undefine_public_key
    MerbRecaptcha::AppHelper.undefine_private_key
  end
  def test_validate_recap_succeeds_without_key_constants_but_with_options
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PUB)
    assert !MerbRecaptcha::AppHelper.const_defined?(:RCC_PRIV)
    assert_nothing_raised do
      @cf.validate_recap({}, {}, :rcc_pub => 'foo', :rcc_priv => 'bar')
    end
  end
  def test_validate_recap_is_correct_with_constants_and_with_options
    # first, with constants
    MerbRecaptcha::AppHelper.define_public_key  # 'foo'
    MerbRecaptcha::AppHelper.define_private_key # 'bar'
    assert @cf.validate_recap({}, {})
    MerbRecaptcha::AppHelper.undefine_public_key
    MerbRecaptcha::AppHelper.undefine_private_key
    # next, with options
    assert @cf.validate_recap({}, {}, :rcc_pub => 'foobar', :rcc_priv => 'blegga')
  end
  
  #
  # unit tests for HTTP/HTTPS-variants of get_captcha() method
  #
  
  def test_get_captcha_uses_http_without_options
    expected= <<-EOF
    <script type=\"text/javascript\" src=\"http://api.recaptcha.net/challenge?k=%s\"> </script>\n      <noscript>\n      <iframe src=\"http://api.recaptcha.net/noscript?k=%s\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    actual = @vf.get_captcha(:rcc_pub => 'foobar', :rcc_priv => 'blegga')
    assert_equal(((expected % ['foobar', 'foobar']).strip), actual.strip)
  end
  def test_get_captcha_uses_https_with_options_true
    expected= <<-EOF
    <script type=\"text/javascript\" src=\"https://api-secure.recaptcha.net/challenge?k=%s\"> </script>\n      <noscript>\n      <iframe src=\"https://api-secure.recaptcha.net/noscript?k=%s\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    actual = @vf.get_captcha(:rcc_pub => 'foobar', :rcc_priv => 'blegga', :ssl => true)
    assert_equal(((expected % ['foobar', 'foobar']).strip), actual.strip)
  end
  def test_get_captcha_uses_http_with_options_false
    expected= <<-EOF
    <script type=\"text/javascript\" src=\"http://api.recaptcha.net/challenge?k=%s\"> </script>\n      <noscript>\n      <iframe src=\"http://api.recaptcha.net/noscript?k=%s\"\n      height=\"300\" width=\"500\" frameborder=\"0\"></iframe><br>\n      <textarea name=\"recaptcha_challenge_field\" rows=\"3\" cols=\"40\">\n      </textarea>\n      <input type=\"hidden\" name=\"recaptcha_response_field\" \n      value=\"manual_challenge\">\n      </noscript>
    EOF
    actual = @vf.get_captcha(:rcc_pub => 'foobar', :rcc_priv => 'blegga', :ssl => false)
    assert_equal(((expected % ['foobar', 'foobar']).strip), actual.strip)
  end
  
private

  def new_client(pubkey='abc', privkey='def', ssl=false)
    MerbRecaptcha::Client.new(pubkey, privkey, ssl)
  end
end


config/hoe.rb
require 'merb_recaptcha/version'

AUTHOR = 'Felix McCoey'  # can also be an array of Authors
EMAIL = "felix.mccoey@gmail.com"
DESCRIPTION = "A Merb adaption of ReCaptcha"
GEM_NAME = 'merb_recaptcha' # what ppl will type to install your gem
RUBYFORGE_PROJECT = 'merb_recaptcha' # The unix name for your project
HOMEPATH = "http://#{RUBYFORGE_PROJECT}.rubyforge.org"
DOWNLOAD_PATH = "http://rubyforge.org/projects/#{RUBYFORGE_PROJECT}"
EXTRA_DEPENDENCIES = [
#  ['activesupport', '>= 1.3.1']
]    # An array of rubygem dependencies [name, version]

@config_file = "~/.rubyforge/user-config.yml"
@config = nil
RUBYFORGE_USERNAME = "unknown"
def rubyforge_username
  unless @config
    begin
      @config = YAML.load(File.read(File.expand_path(@config_file)))
    rescue
      puts <<-EOS
ERROR: No rubyforge config file found: #{@config_file}
Run 'rubyforge setup' to prepare your env for access to Rubyforge
 - See http://newgem.rubyforge.org/rubyforge.html for more details
      EOS
      exit
    end
  end
  RUBYFORGE_USERNAME.replace @config["username"]
end


REV = nil
# UNCOMMENT IF REQUIRED:
# REV = YAML.load(`svn info`)['Revision']
VERS = MerbRecaptcha::VERSION::STRING + (REV ? ".#{REV}" : "")
RDOC_OPTS = ['--quiet', '--title', 'merb_recaptcha documentation',
    "--opname", "index.html",
    "--line-numbers",
    "--main", "README",
    "--inline-source"]

class Hoe
  def extra_deps
    @extra_deps.reject! { |x| Array(x).first == 'hoe' }
    @extra_deps
  end
end

# Generate all the Rake tasks
# Run 'rake -T' to see list of generated tasks (from gem root directory)
$hoe = Hoe.new(GEM_NAME, VERS) do |p|
  p.developer(AUTHOR, EMAIL)
  p.description = DESCRIPTION
  p.summary = DESCRIPTION
  p.url = HOMEPATH
  p.rubyforge_name = RUBYFORGE_PROJECT if RUBYFORGE_PROJECT
  p.test_globs = ["test/**/test_*.rb"]
  p.clean_globs |= ['**/.*.sw?', '*.gem', '.config', '**/.DS_Store']  #An array of file patterns to delete on clean.

  # == Optional
  p.changes = p.paragraphs_of("History.txt", 0..1).join("\n\n")
  #p.extra_deps = EXTRA_DEPENDENCIES

    #p.spec_extras = {}    # A hash of extra values to set in the gemspec.
  end

CHANGES = $hoe.paragraphs_of('History.txt', 0..1).join("\\n\\n")
PATH    = (RUBYFORGE_PROJECT == GEM_NAME) ? RUBYFORGE_PROJECT : "#{RUBYFORGE_PROJECT}/#{GEM_NAME}"
$hoe.remote_rdoc_dir = File.join(PATH.gsub(/^#{RUBYFORGE_PROJECT}\/?/,''), 'rdoc')
$hoe.rsync_args = '-av --delete --ignore-errors'
$hoe.spec.post_install_message = File.open(File.dirname(__FILE__) + "/../PostInstall.txt").read rescue ""


lib/merb_recaptcha.rb
require 'openssl'
require 'base64'
require 'cgi'
require 'net/http'
require 'net/https'

module MerbRecaptcha

  module ViewHelper

    def get_captcha(options={})
      k = MerbRecaptcha::Client.new(options[:rcc_pub] || RCC_PUB, options[:rcc_priv] || RCC_PRIV, options[:ssl] || false)
      r = k.get_challenge(session[:rcc_err] || '', options)
      session[:rcc_err]=''
      r
    end
    def mail_hide(address, contents=nil)
      contents = truncate(address,10) if contents.nil?
      k = MerbRecaptcha::MHClient.new(MH_PUB, MH_PRIV)
      enciphered = k.encrypt(address)
      uri = "http://mailhide.recaptcha.net/d?k=#{MH_PUB}&c=#{enciphered}"
      t =<<-EOF
      <a href="#{uri}"
      onclick="window.open('#{uri}', '', 'toolbar=0,scrollbars=0,location=0,statusbar=0,menubar=0,resizable=0,width=500,height=300'); return false;" title="Reveal this e-mail address">#{contents}</a>
    EOF
    end

  end

  module AppHelper
    private
    def validate_recap(p, errors, options = {})
      rcc=MerbRecaptcha::Client.new(options[:rcc_pub] || RCC_PUB, options[:rcc_priv] || RCC_PRIV)
      res = rcc.validate(request.remote_ip, p[:recaptcha_challenge_field], p[:recaptcha_response_field], errors)
      session[:rcc_err]=rcc.last_error

      res
    end
  end

  class MHClient
    def initialize(pubkey, privkey)
      @pubkey=pubkey
      @privkey=privkey
      @host='mailhide.recaptcha.net'
    end
    def encrypt(string)
      padded = pad(string)
      iv="\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00\00"
      cipher=OpenSSL::Cipher::Cipher.new("AES-128-CBC")
      binkey = @privkey.unpack('a2'*16).map{|x| x.hex}.pack('c'*16)
      cipher.encrypt
      cipher.key=binkey
      cipher.iv=iv
      ciphertext = []
      cipher.padding=0
      ciphertext = cipher.update(padded)
      ciphertext << cipher.final() rescue nil 
      Base64.encode64(ciphertext).strip.gsub(/\+/, '-').gsub(/\//, '_').gsub(/\n/,'')
    end
    def pad(str)
      l= 16-(str.length%16)
      l.times do
        str<< l
      end
      str
    end
  end

  class Client
    def initialize(pubkey, privkey, ssl=false)
      @pubkey = pubkey
      @privkey=privkey
      @host = ssl ? 'api-secure.recaptcha.net':'api.recaptcha.net'
      @vhost = 'api-verify.recaptcha.net'
      @proto = ssl ? 'https' : 'http'
      @ssl = ssl
      @last_error=nil
    end

    def get_challenge(error='', options={})
      s=''
      if options[:options]
        s << "<script type=\"text/javascript\">\nvar RecaptchaOptions = { "
        options[:options].each do |k,v|
          val = (v.class == Fixnum) ? "#{v}" : "\"#{v}\""
          s << "#{k} : #{val}, "
        end
        s.sub!(/, $/, '};')
        s << "\n</script>\n"
      end
      errslug = (error.empty?||error==nil||error=="success") ? '' :  "&error=#{CGI.escape(error)}"
      s <<<<-EOF
      <script type="text/javascript" src="#{@proto}://#{@host}/challenge?k=#{CGI.escape(@pubkey)}#{errslug}"> </script>
      <noscript>
      <iframe src="#{@proto}://#{@host}/noscript?k=#{CGI.escape(@pubkey)}#{errslug}"
      height="300" width="500" frameborder="0"></iframe><br>
      <textarea name="recaptcha_challenge_field" rows="3" cols="40">
      </textarea>
      <input type="hidden" name="recaptcha_response_field" 
      value="manual_challenge">
      </noscript>
      EOF
    end

    def last_error
      @last_error
    end

    def validate(remoteip, challenge, response, errors)
      msg = "Captcha failed."
      return true if  %w(0.0.0.0 127.0.0.1).include?(remoteip)
      unless response and challenge
        errors.add_to_base(msg)
        return false
      end
      proxy_host, proxy_port = nil, nil
      proxy_host, proxy_port = ENV['proxy_host'].split(':')  if ENV.has_key?('proxy_host')
      http = Net::HTTP::Proxy(proxy_host, proxy_port).start(@vhost)
      path='/verify'
      data = "privatekey=#{CGI.escape(@privkey)}&remoteip=#{CGI.escape(remoteip)}&challenge=#{CGI.escape(challenge)}&response=#{CGI.escape(response)}"
      resp, data = http.post(path, data, {'Content-Type'=>'application/x-www-form-urlencoded'})
      response = data.split
      result = response[0].chomp
      @last_error=response[1].chomp
      errors.add_to_base(msg) if  result != 'true'
      result == 'true' 
    end
  end
end
